apache/superset
# fix(chart): Sort By (Backend sorting)

### **SUMMARY**

Currently , if you use the sort by from chart editor control panel it has no effect , the Sort by control attatches an order by to the sql query which actually sorts the data on sql level but not in chart editor . You can see that it gives proper sorted data in sql lab.
The reason for this is that  the post processing applied on the dataframe causes it to lose the sorting applied by the order by query. 
<img width="1496" alt="Screenshot 2024-03-23 at 3 13 24 PM" src="https://github.com/apache/superset/assets/70762780/a028afba-6fb3-4d1b-9a80-d6f44fddafa4">

<img width="1512" alt="Screenshot 2024-03-23 at 3 13 52 PM" src="https://github.com/apache/superset/assets/70762780/67169380-bf5c-4620-b67e-6287cba56643">







Solution : What we basically do is we preserve the order of the dataframe with respect to X axis and we store it in a variable and then re use that variable to re order the dataframe again 

Example : 

In the above screenshots we can see that the output data comes in order "No : Not Willing To " , "Yes willing to" , "No answer" in sql lab if sorted correctly , what we can do is that we can preserve this order in an array as [ "No : Not Willing To " , "Yes willing to" , "No answer"] and we will later use this order to reorder the data after postprocessing so that the data gets correct soring order 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before: 
![screen-capture (11)](https://github.com/apache/superset/assets/70762780/b24b6fc2-a600-40bd-82e1-e180d244ca46)
After: 
![screen-capture (12)](https://github.com/apache/superset/assets/70762780/e86a099a-3854-4dc8-810e-7166058e6d28)

With this fix , can apply sorting on any column available in the dataset, for example we apply sorting by Max(Age), a column which is not applied in the present chart , currently the x axis sort by only allows the applied metrics or columns.. with this fix we can use any column in the dataset

![screen-capture (14)](https://github.com/apache/superset/assets/70762780/b6efa30e-3c87-45ed-bc25-f53dcc01a15e)

Or can write extreme edge cases where the user can write custom sql for ordering data , this is actually the use case in some organisations where the data needs to be sorted in specific order that makes no sense by metric or category name , so currently they make dummy columns and assign them numbers .. this whole process can be avoided if sorting with custom sql can be done :

![Screenshot 2024-03-22 at 6 23 20 PM (2)](https://github.com/apache/superset/assets/70762780/35adbfea-47e7-4332-abb9-c0dc98b7a9ce)

![Screenshot 2024-03-22 at 6 23 37 PM (1)](https://github.com/apache/superset/assets/70762780/891e8ccb-0944-41aa-9cdc-91264b328131)






### TESTING INSTRUCTIONS
Try using Sort By on a chart on Superset 4.0 by any metric and check if the data gets sorted or not, if necessary compare it with sql lab results as it shows the correct order

Then try using sort by ,in chart editor in this fix & try sorting by any metric and compare it with the sql lab results


### ADDITIONAL INFORMATION
